org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/ProjectsManager.java

```diff
@@ -207,7 +212,65 @@ private Set<IPath> removeImportedConfigurations(Set<IPath> configurations, IProj
 			.collect(Collectors.toSet());
 	}
 
-	public void importProjects(IProgressMonitor monitor) {
+	public void importProjects(IProgressMonitor monitor, String... buildFileUris) {
+		if (buildFileUris == null || buildFileUris.length == 0) {
+			importProjectsAutomatically();
+			return;
+		}
+		importProjectsFromSelection(Arrays.asList(buildFileUris));
+	}
+
+	private void importProjectsFromSelection(Collection<String> projectConfigurations) {
```

Comment:
I see you're delegating to the server to calculate the import delta (including toImport projects, toRemove projects, toUpdate projects), that makes the server code a little implicit and complex. I'm wondering whether it's better to let the client to prepare those delta directly?

The server should be as simple as possible, and do whatever the client tells.

Reply:
Let client to resolve the delta is fine.

There is another question: do you think the client should fetch the imported projects and their types before listing the project pickers? (those data will be used to mark the picked flag)

Currently the client just fetches the cached choices and mark them as picked.

Reply:
> There is another question: do you think the client should fetch the imported projects and their types before listing the project pickers? (those data will be used to mark the picked flag)

Yes, it should. That's exactly another comment I want to add in the client PR. The client project picker should mark the imported projects differently. Otherwise, when you click the picker a few times, you will get lost on what projects will be added/removed.

Reply:
> Otherwise, when you click the picker a few times, you will get lost on what projects will be added/removed.

The check flags represent the final state of the imported projects: picked the projects go in and unpicked project go out. Does it really matter which ones are added/removed during the process?

Reply:
For me, I like transparency on what will happen. But looking it at the perspective you presented, it looks fine.